### ISSUE TITLE
Different training loss between runs with identical configuration

### ISSUE THREAD

**Issue by BIGBALLON** [pytorch/torchtitan#1720]
I noticed that when training dense models (``llama3-8b`` and the ``debug_model`` of llama3) using:

- the same TOML configuration file
- the same GPUs, and
- the same launch command

the resulting training loss curves are not identical.

Here is an example from TensorBoard:

<img width="1424" height="705" alt="Image" src="https://github.com/user-attachments/assets/c1b17b4b-b803-41b1-90e1-c0a8b2f2e772" />

<img width="1559" height="572" alt="Image" src="https://github.com/user-attachments/assets/ca241ef6-20f4-48b8-ad8b-c5a3c4928c09" />

Could you help explain what might be causing this difference? Some possible reasons I am aware of:

- Random initialization or stochastic data loading order
- Non-deterministic CUDA kernels (e.g., atomic operations, reduction ops)
- Distributed communication scheduling introducing slight variations

Is this expected behavior, or should the runs be bitwise reproducible under the same setup?

Thanks!

**Comment by tianyu-l** [comment]
> Random initialization or stochastic data loading order

Data loading is deterministic
Model init is not (you can turn on `--training.seed`)

> Non-deterministic CUDA kernels (e.g., atomic operations, reduction ops)

you can try `--training.deterministic`, but this flag may not work with all ops in torchtitan

> Distributed communication scheduling introducing slight variations

This one I'm not sure. cc @ngimel
